### ISSUE TITLE
DefaultResolveType performance

### ISSUE THREAD

**Issue by lpagithub** [OData/odata.net#2458]
Using generated client code where DefaultResolveType is called and if I instantiate DataServiceContext before some method call performance is bad. It seems because of resolveTypesCache is not static.

### Assemblies affected

Microsoft.OData.Client 7.12

### Reproduce steps

Run loop where DataServiceContext is instantiated and call some query.

### Expected result

Time to execute some function should be close to download response.

### Actual result

Time is 5 time bigger.

**Comment by corranrogue9** [comment]
Hey @lpagithub, can you provide a more specific example? I want to make sure I have a proper comparison with the use of `DefaultResolveType` and whatever other code is 5 times faster.

**Comment by lpagithub** [comment]
Hi @corranrogue9, I will try. 

As mentioned, generated reference client is used (OData Connected Service, VS 2019).

In that code it is method:
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.OData.Client.Design.T4", "#VersionNumber#")]
        protected global::System.Type ResolveTypeFromName(string typeName)
        {
            global::System.Type resolvedType = this.DefaultResolveType(typeName, "Default", "SomeNs");
            if ((resolvedType != null))
            {
                return resolvedType;
            }
            return null;
        }

Here is our very basic code:
public ConsentContent GetVersion(string consentId)
        {
            var container = CreateContainer();
            var sw = Stopwatch.StartNew();
            var specificConsentContent = container.ConsentContent.ByKey(new Guid(consentId)).GetValue();
            Log.Info($"GetVersion: {sw.ElapsedMilliseconds}", this);
            return Map(specificConsentContent);
        }

protected Container CreateContainer()
        {
            var container = new Container(new Uri(ConfigurationManager.AppSettings["SomeApiAddress"]));
            container.BuildingRequest += (sender, eventArgs) =>
            {
                eventArgs.Headers.Add("Ocp-Apim-Subscription-Key", ConfigurationManager.AppSettings["SomeKey"]);
                eventArgs.Headers.Add("No-Redirect", "true");
            };
            return container;
        }

Log:
GetVersion: 288
GetVersion: 279
GetVersion: 274
GetVersion: 382

Now with small hack in Reference file using the same code from DataServiceContext, but with 

private **static** ConcurrentDictionary<string, Type> resolveTypesCache = new ConcurrentDictionary<string, Type>();

global::System.Type resolvedType = this.DefaultResolveType2(typeName, "Default", "SomeNs");
DefaultResolveType2 is copy from DataServiceContext

Log:
GetVersion: 57
GetVersion: 68
GetVersion: 34
GetVersion: 69

It is test environment so times are bit unstable and depends also on response size.

Of course every time DataServiceContext is instantiated loosing resolveTypesCache if not static and because we have lot of assemblies performance is bad in TryResolveType. I could actually see the issue in VS "Performance Profiler".

Hope it helps and I'm not mistaking. Thanks btw.

**Comment by corranrogue9** [comment]
Is there a reason that you are instantiating multiple `DataServiceContext`s? Can you share what that code looks like?

**Comment by lpagithub** [comment]
Yes, I was thinking also about new constructor param, but so proxy generation should be modified, or resolve type virtual method, where in this case types can be resolved using Assembly.GetExecutingAssembly() and probably cached again.
But, not sure I understand `DefaultResolveType` and `ClientTypeUtil.TryResolveType` methods, because server namespace (full ns) is stripped in the beginning of TryResolveType and only client (language depending) ns is used. So, there is no way to return something else when another service is used. And, `resolveTypesCache` is based on server namespace so it should not be conficts within cache. Will you also check the code? Maybe I'm missing something.

**Comment by gathogojr** [comment]
Fixed in https://github.com/OData/odata.net/pull/2569
